### PR TITLE
fix(auth): bind DCR client to user in production Google callback (#648)

### LINKED ISSUE #648: Production OAuth callback never binds DCR client to a user account — list_memories & summarize_context fail in prod
### Area

OAuth / authentication

### What happened?

In production, the MCP tools `list_memories` and `summarize_context` always fail for real (Google-authenticated) clients with:

```
Client is not associated with a user account; per-user memory scoping is required.
```

(`request_id` examples from a live e2e run: `req_011CbzMf4QEKtWymeygrJxBw`, `req_011CbzMv99A59HqMss5ReMBP`.)

### Root cause

MCP access tokens carry only `client_id` + `scope` — no user identity (`create_token_pair`, `src/hive/storage.py`). The tools resolve identity by reading `client.owner_user_id` off the persisted client record (`_auth` → `get_client`, `src/hive/server.py`), and both `list_memories` and `summarize_context` hard-reject when it is `None` (`server.py` ~L1226 and ~L1336).

`client.owner_user_id` is set in exactly **one** place: `_bypass_associate_user` (`src/hive/auth/oauth.py:134`), which runs **only** under `HIVE_BYPASS_GOOGLE_AUTH`. The production `google_callback` (`src/hive/auth/oauth.py:253`) verifies the Google identity, issues the auth code, and redirects — but never upserts a `User` and never sets `client.owner_user_id`. The token endpoint doesn't either. So in prod `owner_user_id` is always `None` and these two tools are permanently broken.

**Why CI never caught it:** e2e runs in bypass mode, which is the one path that *does* the binding — tests pass while prod is broken. Classic test/prod path divergence.

### Steps to reproduce

1. Connect a fresh MCP client to `https://hive.warlordofmars.net/mcp` via the real Google OAuth flow (not bypass).
2. Call `list_memories(tag="anything")` → fails with the user-account error.
3. Same for `summarize_context(topic="anything")`.

### Expected behaviour

A client that completes the production Google OAuth flow is associated with the authenticated user, and `list_memories` / `summarize_context` work.

### Proposed fix

- In `google_callback`, replicate what `_bypass_associate_user` already does: upsert a `User` from the verified Google claims and set `client.owner_user_id = user.user_id` (persist with `put_client`). Bind at the point a real user identity is established.
- Decide where binding belongs if a client is reused across users (today `_bypass_associate_user` only binds when `owner_user_id is None` — keep that first-bind-wins semantics or revisit).
- Add an **integration test that exercises the real callback path** (not bypass) and asserts `owner_user_id` is set and that `list_memories` / `summarize_context` succeed afterwards.

### Follow-up

Once unblocked, re-test `summarize_context` output quality — the live tester reported a suspected concatenation-vs-synthesis issue that couldn't be reached because the tool errors before producing output.

### Environment

- Diagnosed against prod 2026-06-12; confirmed live (a `list_memories` call from a real client still fails today).
- Touches `auth/` → not `agent-safe`.

### PR DESCRIPTION
**Pull request by warlordofmars**
Closes #648

## Summary

In production the MCP tools `list_memories` and `summarize_context` always failed for real Google-authenticated clients with *"Client is not associated with a user account; per-user memory scoping is required."*

The tools resolve identity from `client.owner_user_id`, but that field was set in exactly **one** place — the `HIVE_BYPASS_GOOGLE_AUTH` shortcut (`_bypass_associate_user`). The production `google_callback` verified the Google identity and issued the auth code but **never** upserted a `User` or bound the client. CI never caught it because e2e runs bypass-only — the one path that *does* bind — so tests passed while prod was broken. Classic test/prod path divergence.

## Fix

- Renamed `_bypass_associate_user` → `_associate_user_with_client` and route **both** the production callback and the bypass shortcut through it, so the two paths bind users **identically** (eliminating the exact divergence that caused the bug).
- In `google_callback`, call the helper right after the verified-email + allowlist checks and before issuing the auth code.

## Decisions (please veto if you disagree)

- **First-bind-wins** — `owner_user_id` is set only when currently `None`; a client already owned by one user is never re-pointed at a different account that later authenticates through it. This matches the pre-existing bypass behaviour. The alternative (rebind-on-each-login) would let a later user silently take over an existing client, so first-bind-wins is the safer default. Happy to revisit if you'd prefer rebind semantics.
- **Scope = `owner_user_id` only** — `OAuthClient.workspace_id` is intentionally **not** set here; it stays `None` until the #490 workspaces migration / #491 DCR tightening. Mirrors the bypass path and keeps this p0 out of the in-flight workspaces work.
- **`display_name` from the email local-part** (mirrors the bypass path), not the Google `name` claim — see *Known / deferred*.

## Tests (TDD — both watched fail first, for the issue's exact reason)

- **Unit** (`tests/unit/test_auth.py`): `test_success_binds_authenticated_user_to_client`, `test_success_binds_user_only_once`.
- **Integration** (`tests/integration/test_oauth_user_binding.py`): drives the **real** `google_callback` (mocking *only* Google's network calls — `exchange_google_code` / `verify_google_id_token` / allowlist), then asserts `owner_user_id` is bound to the upserted user **and** that `remember` → `list_memories` → `summarize_context` all succeed against DynamoDB Local. This is the regression guard the bug lacked, because e2e only ever exercised the bypass path.

### Verification

- Combined unit + integration: **920 passed, 19 skipped, 100% coverage** (`auth/oauth.py` 199 stmts, 0 missed).
- Frontend vitest **887 passed**; ruff / mypy / copyright headers all green.
- Adversarial multi-lens review (correctness / auth-security / test-fidelity / scope): **0 confirmed findings**. One reviewer independently reverted the binding line and confirmed both unit tests fail with `assert None is not None`, then restored it — confirming the tests genuinely guard the regression.

## Known / deferred (out of scope for this p0)

`display_name` source-of-truth differs between this MCP-callback upsert and the management-UI login upsert (`mgmt_auth._upsert_user`); both already overwrite on every login (pre-existing, not introduced here). A Google-then-MCP login sequence would downgrade e.g. "John Carter" → "johncarter". Candidate for a separate low-priority cleanup (set `display_name` only on create, or share one helper across both paths).

## Live verification

A true prod check requires the post-merge prod deploy plus a **real Google browser login**, which can't be driven headlessly from CI. The integration test is the faithful local reproduction of the exact failing path (fails before this change, passes after). Recommend confirming live after the prod deploy with the same client that filed the issue (the issue cites live request IDs).

## Merge

Touches `auth/` → **not `agent-safe`**. Copilot review will run, but auto-merge is intentionally **not** armed — please review and merge manually.

### CODE REVIEW

File: src/hive/auth/oauth.py [hunk 1]
```diff
@@ -290,6 +297,13 @@ async def google_callback(
     if not is_email_allowed(email):
         raise HTTPException(status_code=403, detail=f"Email {email!r} is not authorised")
 
+    # Bind the DCR client to the now-verified user (first-bind-wins) so that
+    # user-scoped MCP tools work for real Google-authenticated clients. Without
+    # this, production clients kept owner_user_id=None and list_memories /
+    # summarize_context always failed — the bug behind #648. Mirrors the
+    # HIVE_BYPASS_GOOGLE_AUTH path via the same shared helper.
+    _associate_user_with_client(storage, pending.client_id, email)
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/auth/oauth.py:357]: Fixed in 2c37865. `_associate_user_with_client` is now single-user: once a client is bound, a different verified account is rejected with **403** *before any write* (first-time binding and same-user re-login are unchanged). Both the production callback and the bypass shortcut go through this shared helper, so the guard applies to both. This was a real cross-user exposure that first-bind-wins alone didn't close — thanks for catching it.

File: tests/unit/test_auth.py [hunk 2]
```diff
@@ -633,6 +633,70 @@ def test_google_token_error_returns_400(self, oauth_client):
...
+                )
+
+        assert _callback_as("first@example.com").status_code == 302
+        first_owner = storage.get_client(client.client_id).owner_user_id
+        assert first_owner is not None
+
+        assert _callback_as("second@example.com").status_code == 302
+        # Binding stays with the first user even though a different account
+        # later authenticated through the same client.
+        assert storage.get_client(client.client_id).owner_user_id == first_owner
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply tests/unit/test_auth.py]: Fixed in 2c37865. The test now asserts the second *different* account receives **403** with the original `owner_user_id` left unchanged (renamed `test_different_user_on_owned_client_rejected`), and a new `test_same_user_relogin_keeps_binding` confirms the owner can still re-authenticate. An integration test (`test_callback_rejects_second_user_on_owned_client`) covers the same boundary against DynamoDB Local.

File: src/hive/auth/oauth.py [hunk 3]
```diff
@@ -147,6 +162,21 @@ def _bypass_associate_user(storage: HiveStorage, client_id: str, email: str) ->
     role = "admin" if is_admin_email(email) else "user"
 
     user = storage.get_user_by_email(email)
+    client = storage.get_client(client_id)
+
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/auth/oauth.py:168]: Fixed in a9769af. `_associate_user_with_client` now raises 400 ("Unknown or deleted client") as soon as `get_client` returns None — before any user upsert — so a client deleted between /oauth/authorize and the callback yields no orphan user and no auth code. Added `test_callback_with_deleted_client_fails_without_side_effects`. Good catch.

File: src/hive/auth/oauth.py [hunk 4]
```diff
@@ -155,8 +187,7 @@ def _bypass_associate_user(storage: HiveStorage, client_id: str, email: str) ->
     user.last_login_at = now
     storage.put_user(user)
 
-    client = storage.get_client(client_id)
-    if client is not None and client.owner_user_id is None:
+    if client.owner_user_id is None:
         client.owner_user_id = user.user_id
         storage.put_client(client)
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/auth/oauth.py]: Fixed in 03b4394. Added `HiveStorage.bind_client_owner` — a conditional `UpdateItem` (`attribute_exists(PK) AND attribute_not_exists(owner_user_id)`) that returns whether this caller won the bind, mirroring `mark_auth_code_used`. The callback claims ownership through it; a race-loser re-reads and gets 403 unless the winner is the same user, and `put_user` runs only after the bind settles so a rejected login persists nothing. Added storage-layer tests + a callback race-loss regression test. Spot-on — this defeated both first-bind-wins and the 403 guard under concurrency.

File: tests/integration/test_oauth_user_binding.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+def _make_context(token_str: str):
+    """Build a minimal MCP Context carrying an Authorization header."""
+    ctx = MagicMock()
+    ctx.request_context = MagicMock()
+    ctx.request_context.meta = {"Authorization": f"Bearer {token_str}"}
+    return ctx
+
+
+@pytest.fixture()
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply tests/integration/test_oauth_user_binding.py]: Done in 03b4394 — `google_oauth_setup` is now module-scoped to match the other integration fixtures in this file; the two tests use distinct clients/users/tags so they don't collide on the shared table.

File: src/hive/auth/oauth.py [hunk 6]
```diff
@@ -147,18 +164,44 @@ def _bypass_associate_user(storage: HiveStorage, client_id: str, email: str) ->
...
+    # concurrently, exactly one bind wins. The loser re-reads and is refused
+    # unless the winner happens to be this same user. put_user runs only after
+    # the bind is settled, so a rejected login persists nothing.
+    if client.owner_user_id is None and not storage.bind_client_owner(client_id, user.user_id):
+        winner = storage.get_client(client_id)
+        if winner is None or winner.owner_user_id != user.user_id:
+            raise HTTPException(
+                status_code=403,
+                detail="This client is already associated with a different user account.",
+            )
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/auth/oauth.py:215]: Fixed in bdbd92d. The bind-race loser path now: (1) returns **400** ("Unknown or deleted client") when the re-read is None — the client was deleted mid-race; (2) resolves by **identity** — it looks up the winner's owner and, if that owner's email matches the verified email, treats it as the same account and proceeds *without* writing a duplicate user (fixing the spurious 403 on two concurrent first-logins by the same account); a genuinely different owner still gets 403. Added two race regression tests. Thanks — both edge cases were real.

File: src/hive/auth/oauth.py [hunk 7]
```diff
@@ -147,18 +164,53 @@ def _bypass_associate_user(storage: HiveStorage, client_id: str, email: str) ->
...
+        raise HTTPException(status_code=400, detail="Unknown or deleted client")
+
+    # Single-user boundary: refuse to proceed when the client is already bound
+    # to someone other than the authenticating user. Checked before any write so
+    # a rejected login leaves no trace.
+    if client.owner_user_id is not None and (user is None or user.user_id != client.owner_user_id):
+        raise HTTPException(
+            status_code=403,
+            detail="This client is already associated with a different user account.",
+        )
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/auth/oauth.py:189]: Fixed in 7aef1c5. The single-user check now distinguishes a dangling binding from an intruder: it looks up the existing owner via `get_user_by_id`, and if that record is gone returns **400** ("bound to a user account that no longer exists") instead of the misleading 403; a genuine different-user attempt still gets 403. Added a regression test (admin-deleted owner). Good eye.
